cartridge/shop/managers.py

```diff
@@ -22,26 +22,20 @@ def from_request(self, request):
         cart.
         """
         cart_id = request.session.get("cart", None)
-        cart = None
-        if cart_id:
+        cart = self.current().filter(id=cart_id)
+        last_updated = now()
+        # Update timestamp and clear out old carts.
+        if cart_id and cart.update(last_updated=last_updated):
+            self.expired().delete()
+        else:
             try:
-                cart = self.current().get(id=cart_id)
-            except self.model.DoesNotExist:
                 request.session["cart"] = None
-            else:
-                # Update timestamp and clear out old carts.
-                cart.last_updated = now()
-                cart.save()
-                self.expired().delete()
-        if not cart:
-            # Forget what checkout step we were up to.
-            try:
+                # Forget what checkout step we were up to.
                 del request.session["order"]["step"]
                 request.session.modified = True
             except KeyError:
                 pass
-            cart = self.model(last_updated=now())
-        return cart
+        return self.model(id=cart_id, last_updated=last_updated)
```

Comment:
How about `assert len(Cart._meta.fields) == 2`? That should pop up during testing immediately if it gets broken.

DB call overhead increases quite a bit when the DB and the web server aren't sitting close together, so that one call could add a whole roundtrip (50ms+) in those situations. I'm trying to cut down on spurious DB dips (you may have noticed from my recent pull request rampage) to make it more realistic to put a second web server to service Europe and get faster page load times for European visitors. Caching can go a long way, but I'm already not a fan of the fact that the cart needs several round trips with the DB on every request.